### DISCUSSION TITLE
is it possible to add random 16 or 256 colors to my prompt?

### DISCUSSION THREAD

**Discussion by junguler** [fish-shell/fish-shell#10783] [Q&A] [answered]
in zsh i could do `%F{$(($RANDOM%6+1))} %D{%H:%M} %f` and in bash `$(tput setaf $((RANDOM%6+1))) \A $(tput sgr0)` but i can't find a similar solution for fish

i've been using perplexity to answer small questions like this but it fails miserably when it comes to customizing prompts, no matter which shell or extra prompt i want to get information on

searching the internet proper didn't yield any results either, i couldn't even get starship to use random colors, is this impossible or i'm just looking at the wrong places?

here is an example of what i mean by random colors

![random](https://github.com/user-attachments/assets/0489d296-c8ea-4040-b332-8de61c1072c6)

**Comment by faho** [comment, marked as answer]
Use the [random builtin](https://fishshell.com/docs/current/cmds/random.html).

For the 16 colors, you can do

`set_color (random choice (set_color --print-colors))`

You probably want to ignore colors that look bad on your background, e.g. via `string match -v`, as [I've done in my color randomizer thing](https://github.com/faho/fish-snippets/blob/973afe8e31a02eeada23d29fc4b53456eda18745/conf.d/goblin-mode.fish#L7).

set_color has no support for specifying a 256-palette color, so you'd have to create the escape sequence yourself, either via printf or via `tput` as you've done for bash.

Or you could create an RGB color string and use that.

> i've been using perplexity to answer small question

Yes, AI tools are in fact extra terrible at fish, and will mislead you, and will waste our time. I recommend not using them.

> i couldn't even get starship to use random colors

starship is an external tool that prints whatever it wants. you'll have to see how to tell starship that it should print random colors.

**Reply by junguler** [reply]
thank you for the great and informative answer you gave, i really appreciate that

**Reply by junguler** [reply]
hello again, after messing around with my prompt for a few hours i have a few other questions if you are willing to answer

is it possible to exclude more than one color when using the 16 colors? basically i don't want (black brblack white normal) to be included 

also i would love to know a little more about defining my own colors and how would i go about using them randomly in the prompt, a small example would help a lot as i am quite new to how fish does things

**Reply by faho** [reply]
> is it possible to exclude more than one color when using the 16 colors? basically i don't want (black brblack white normal) to be included

If you want a specific set of colors, it's probably simplest to just give the list to `random choice`. It just takes a number of arguments and randomly picks one, so:

```fish
random choice blue brblue brcyan brgreen brmagenta brred brwhite bryellow cyan green magenta red yellow
```

> defining my own colors 

`set_color` takes one of the 16 named colors or an RGB color code (six hex digits like "aa4422")

> using them randomly in the prompt

```fish
function fish_prompt
    set -l random_color (random choice blue brblue brcyan brgreen brmagenta brred brwhite bryellow cyan green magenta red yellow)
    echo (set_color $random_color)$PWD (set_color normal)'>'
end
```

See https://fishshell.com/docs/current/prompt.html for more on how to write a prompt.

Or see [the disco prompt](https://github.com/fish-shell/fish-shell/blob/master/share/tools/web_config/sample_prompts/disco.fish), which turns the working directory into a color. So it changes color when you change directories, but always the same color for the same directory.

**Reply by junguler** [reply]
thanks again, this is exactly what i wanted
